pkg/riff/commands/credential_set.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	if opts.DockerHubId != "" && len(opts.DockerHubPassword) == 0 {
+		errs = errs.Also(cli.ErrMissingField("<docker-hub-password>"))
+	}
+
+	if len(opts.RegistryPassword) != 0 && opts.RegistryUser == "" {
+		errs = errs.Also(cli.ErrMissingField(cli.RegistryUserFlagName))
+	}
+
+	if opts.SetDefaultImagePrefix && opts.Registry != "" {
+		errs = errs.Also(cli.ErrInvalidValue("cannot be used with registry", cli.SetDefaultImagePrefixFlagName))
+	}
+
+	return errs
+}
+
+func NewCredentialSetCommand(c *cli.Config) *cobra.Command {
+	opts := &CredentialSetOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "set",
+		Short:   "<todo>",
+		Example: "<todo>",
+		Args: cli.Args(
+			cli.NameArg(&opts.Name),
+		),
+		PreRunE: cli.Sequence(
+			func(cmd *cobra.Command, args []string) error {
+				if opts.DockerHubId != "" {
+					if err := cli.ReadStdin(c, &opts.DockerHubPassword, "Docker Hub password")(cmd, args); err != nil {
+						return err
+					}
+				}
+				if opts.RegistryUser != "" {
+					if err := cli.ReadStdin(c, &opts.RegistryPassword, "Registry password")(cmd, args); err != nil {
+						return err
+					}
+				}
+				return nil
+			},
+			cli.ValidateOptions(opts),
+		),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// get desired credential and image prefix
+			secret, defaultImagePrefix, err := makeCredential(opts)
+			if err != nil {
+				return err
+			}
+
+			if err := setCredential(c, opts, secret); err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "Set credentials %q\n", opts.Name)
+
+			if opts.SetDefaultImagePrefix {
+				if defaultImagePrefix == "" {
+					fmt.Fprintf(cmd.OutOrStdout(), "Unable to derive default image prefix\n")
+				} else {
+					err := setDefaultImagePrefix(c, opts, defaultImagePrefix)
+					if err != nil {
+						return err
+					}
+					fmt.Fprintf(cmd.OutOrStdout(), "Set default image prefix to %q\n", defaultImagePrefix)
+				}
+			}
+
+			return nil
+		},
+	}
+
+	cli.NamespaceFlag(cmd, c, &opts.Namespace)
+	cmd.Flags().StringVar(&opts.DockerHubId, cli.StripDash(cli.DockerHubFlagName), "", "<todo>")
+	cmd.Flags().StringVar(&opts.GcrTokenPath, cli.StripDash(cli.GcrFlagName), "", "<todo>")
+	cmd.Flags().StringVar(&opts.Registry, cli.StripDash(cli.RegistryFlagName), "", "<todo>")
+	cmd.Flags().StringVar(&opts.RegistryUser, cli.StripDash(cli.RegistryUserFlagName), "", "<todo>")
+	cmd.Flags().BoolVar(&opts.SetDefaultImagePrefix, cli.StripDash(cli.SetDefaultImagePrefixFlagName), false, "<todo>")
+
+	return cmd
+}
+
+func makeCredential(opts *CredentialSetOptions) (*corev1.Secret, string, error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: opts.Namespace,
+			Name:      opts.Name,
+			Labels: map[string]string{
+				build.CredentialLabelKey: "",
+			},
+		},
+	}
+	defaultPrefix := ""
+
+	switch {
+	case opts.DockerHubId != "":
+		secret.Annotations = map[string]string{
+			"build.knative.dev/docker-0": "https://index.docker.io/v1/",
```

Comment:
Why are we using a knative/build annotation rather than our own?

Reply:
These annotations are how Knative Build converts the secret into something that can be consumed by docker. See https://knative.dev/docs/build/auth/

If we used our own annotation, or the equivalent, then we'd need the controller to update the secret to add/update the knative annotation.